parsers/OPENNEM.py

```diff
@@ -80,181 +81,267 @@ def dataset_to_df(dataset):
         interval += "in"
 
     index = pd.date_range(start=dt_start, end=dt_end, freq=interval)
-    assert len(index) == len(series['data'])
-    df = pd.DataFrame(index=index, data=series['data'], columns=[name])
+    assert len(index) == len(series["data"])
+    df = pd.DataFrame(index=index, data=series["data"], columns=[name])
+
     return df
 
 
 def generate_url(zone_key, is_flow, target_datetime, logger) -> str:
     if target_datetime:
         network = ZONE_KEY_TO_NETWORK[zone_key]
         # We will fetch since the beginning of the current month
-        month = arrow.get(target_datetime).floor('month').format('YYYY-MM-DD')
+        month = arrow.get(target_datetime).floor("month").format("YYYY-MM-DD")
         if is_flow:
-            url = f'http://api.opennem.org.au/stats/flow/network/{network}?month={month}'
+            url = (
+                f"http://api.opennem.org.au/stats/flow/network/{network}?month={month}"
+            )
         else:
             region = ZONE_KEY_TO_REGION.get(zone_key)
-            url = f'http://api.opennem.org.au/stats/power/network/fueltech/{network}/{region}?month={month}'
+            url = f"http://api.opennem.org.au/stats/power/network/fueltech/{network}/{region}?month={month}"
     else:
         # Contains flows and production combined
-        url = f'https://data.opennem.org.au/v3/clients/em/latest.json'
+        url = f"https://data.opennem.org.au/v3/clients/em/latest.json"
 
     return url
 
 
-def fetch_main_df(data_type, zone_key=None, sorted_zone_keys=None, session=None, target_datetime=None, logger=logging.getLogger(__name__)):
+def fetch_main_df(
+    data_type,
+    zone_key=None,
+    sorted_zone_keys=None,
+    session=None,
+    target_datetime=None,
+    logger=logging.getLogger(__name__),
+):
     region = ZONE_KEY_TO_REGION.get(zone_key)
     url = generate_url(
         zone_key=zone_key or sorted_zone_keys[0],
         is_flow=sorted_zone_keys is not None,
         target_datetime=target_datetime,
-        logger=logger)
+        logger=logger,
+    )
 
     # Fetches the last week of data
-    logger.info(f'Requesting {url}..')
+    logger.info(f"Requesting {url}..")
     r = (session or requests).get(url)
     r.raise_for_status()
-    logger.debug('Parsing JSON..')
-    datasets = r.json()['data']
-    logger.debug('Filtering datasets..')
-    filtered_datasets = [
-        ds for ds in datasets
-        if ds['type'] == data_type and (
-            (zone_key and ds.get('region') == region)
-            or (sorted_zone_keys and ds.get('id').split('.')[-2] == EXCHANGE_MAPPING_DICTIONARY['->'.join(sorted_zone_keys)]['region_id'])
-        )
-    ]
-    logger.debug('Concatenating datasets..')
+    logger.debug("Parsing JSON..")
+    datasets = r.json()["data"]
+    logger.debug("Filtering datasets..")
+
+    def filter_dataset(ds: dict) -> bool:
```

Comment:
Extracted this function outside of list comprehension to make code easier to read